### PR TITLE
DIRECTOR: Add Lingo v4 bytecode interpreter

### PR DESCRIPTION
**Pull request by moralrecordings**
This patch adds the foundations of a bytecode interpreter for Director v4 Lingo scripts, using the same execution framework as the Lingo parser. 

There's still quite a lot of work to be done, especially with respect to context management, script invocation, event callbacks, etc.

### CODE REVIEW

File: engines/director/lingo/lingo-bytecode.cpp [hunk 1]
```diff
@@ -0,0 +1,311 @@
...
+	Common::Array<Datum> constsData;
+
+	// read each entry in the reference table.
+	stream.seek(constsOffset);
+	for (uint16 i=0; i < constsCount; i++) {
+		Datum constant;
+		uint16 constType = stream.readUint16();
+		uint32 value = stream.readUint32();
+		switch (constType) {
+			case 1: // String type
```
Comment (sev-) [inline engines/director/lingo/lingo-bytecode.cpp]: Indentation of `case` statements must be aligned with their `switch` counterparts as per ou code formatting conventions.

File: engines/director/lingo/lingo-bytecode.cpp [hunk 2]
```diff
@@ -0,0 +1,311 @@
...
+
+	// copy the storage area first.
+	uint32 codeStoreSize = functionsOffset - codeStoreOffset;
+	stream.seek(codeStoreOffset);
+	byte *codeStore = (byte *)malloc(codeStoreSize);
+	stream.read(codeStore, codeStoreSize);
+
+	// read each entry in the function table.
+	stream.seek(functionsOffset);
+	for (uint16 i=0; i<functionsCount; i++) {
```
Comment (sev-) [inline engines/director/lingo/lingo-bytecode.cpp]: whitespaces.

File: engines/director/lingo/lingo-bytecode.cpp [hunk 3]
```diff
@@ -0,0 +1,311 @@
...
+		while (pointer < startOffset + length - codeStoreOffset) {
+			uint8 opcode = codeStore[pointer];
+
+			pointer += 1;
+			if (_lingoV4.contains(opcode)) {
+				offsetList.push_back(_currentScript->size());
+				g_lingo->code1(_lingoV4[opcode]->func);
+				size_t argc = strlen(_lingoV4[opcode]->proto);
+				for (uint c=0; c < argc; c++) {
+					switch (_lingoV4[opcode]->proto[c]) {
```
Comment (sev-) [inline engines/director/lingo/lingo-bytecode.cpp]: Another over-indented set of `case` statements.

File: engines/director/lingo/lingo-bytecode.cpp [hunk 4]
```diff
@@ -0,0 +1,311 @@
...
+						default:
+							break;
+					}
+				}
+
+			} else {
+				// unimplemented instruction
+				if (opcode < 0x40) { // 1 byte instruction
+					offsetList.push_back(_currentScript->size());
+					g_lingo->code1(Lingo::c_unk);
```
Comment (sev-) [inline engines/director/lingo/lingo-bytecode.cpp]: Should we perhaps print out the codes of the unimplemented instrucitons? Then we could see what is missing in the traces.

File: engines/director/lingo/lingo-bytecode.cpp [hunk 5]
```diff
@@ -0,0 +1,311 @@
...
+ */
+
+#include "director/lingo/lingo.h"
+
+namespace Director {
+
+static struct LingoV4Bytecode {
+	const uint8 opcode;
+	const inst func;
+	const char *proto;
```
Comment (sev-) [inline engines/director/lingo/lingo-bytecode.cpp]: It looks that the number of arguments depends on the opcode. Is the `proto` really needed here?

File: engines/director/lingo/lingo-bytecode.cpp [hunk 6]
```diff
@@ -0,0 +1,311 @@
...
+
+		uint16 nameIndex = stream.readUint16();
+		stream.readUint16();
+		uint32 length = stream.readUint32();
+		uint32 startOffset = stream.readUint32();
+		uint16 argCount = stream.readUint16();
+		uint32 argOffset = stream.readUint32();
+		uint16 varCount = stream.readUint16();
+		uint32 varNamesOffset = stream.readUint32();
+		stream.readUint16();
```
Comment (sev-) [inline engines/director/lingo/lingo-bytecode.cpp]: Would be beneficial to print these out on debug level 8+, so we could potentially figure out their meaning.

File: engines/director/lingo/lingo-code.cpp [hunk 7]
```diff
@@ -85,6 +86,8 @@ static struct FuncDescr {
 	{ Lingo::c_starts,		"c_starts",		"" },
 	{ Lingo::c_intersects,	"c_intersects",	"" },
 	{ Lingo::c_within,		"c_within",		"" },
+	{ Lingo::c_field,       "c_field",      "" },
+	{ Lingo::c_of,		    "c_of",		    "" },	// D9
```
Comment (sev-) [inline engines/director/lingo/lingo-code.cpp]: D9, as Director 9?

File: engines/director/lingo/lingo-code.cpp [hunk 8]
```diff
@@ -57,11 +57,12 @@ static struct FuncDescr {
 } funcDescr[] = {
 	{ 0,					"STOP",			"" },
 	{ Lingo::c_xpop,		"c_xpop",		"" },
-	{ Lingo::c_arraypush,	"c_arraypush",		"i" },
+	{ Lingo::c_argspush,	"c_argspush",	"i" },
+	{ Lingo::c_arraypush,	"c_arraypush",	"i" },
 	{ Lingo::c_printtop,	"c_printtop",	"" },
-	{ Lingo::c_constpush,	"c_constpush",	"i" },
+	{ Lingo::c_intpush,	"c_intpush",	"i" },
```
Comment (sev-) [inline engines/director/lingo/lingo-code.cpp]: What is the purpose of renaming `c_constpush` and `c_fconstpush`?

File: engines/director/lingo/lingo-code.cpp [hunk 9]
```diff
@@ -196,8 +205,8 @@ void Lingo::c_printtop(void) {
 	}
 }
 
-void Lingo::c_constpush() {
-	Datum d;
+void Lingo::c_intpush() {
+    Datum d;
```
Comment (sev-) [inline engines/director/lingo/lingo-code.cpp]: Indentation is wrong here

File: engines/director/lingo/lingo-codegen.cpp [hunk 10]
```diff
@@ -274,22 +274,20 @@ int Lingo::codeFloat(double f) {
 	return _currentScript->size();
 }
 
-int Lingo::codeConst(int val) {
-	int res = g_lingo->code1(g_lingo->c_constpush);
+int Lingo::codeInt(int val) {
```
Comment (sev-) [inline engines/director/lingo/lingo-codegen.cpp]: Why did you change the implementation?

File: engines/director/lingo/lingo-gr.y [hunk 11]
```diff
@@ -150,7 +150,8 @@ asgn: tPUT expr tINTO ID 		{
 		$$ = $4;
 		delete $2; }
 	| tSET THEENTITY '=' expr	{
-		g_lingo->codeConst(0); // Put dummy id
+		g_lingo->code1(g_lingo->c_intpush);
```
Comment (sev-) [inline engines/director/lingo/lingo-gr.y]: This change is not clear.

File: engines/director/score.cpp [hunk 12]
```diff
@@ -180,6 +180,30 @@ void Score::loadArchive() {
 	setSpriteCasts();
 	loadSpriteImages(false);
 
+	// Try to load script name lists
+	if (_vm->getVersion() == 4) {
```
Comment (sev-) [inline engines/director/score.cpp]: Is it only D4? or D4+?

File: engines/director/score.cpp [hunk 13]
```diff
@@ -842,6 +896,23 @@ bool Score::processImmediateFrameScript(Common::String s, int id) {
 	return false;
 }
 
+void Score::loadLingoNames(Common::SeekableSubReadStreamEndian &stream) {
+	if (_vm->getVersion() == 4) {
```
Comment (sev-) [inline engines/director/score.cpp]: Same question, is it only D4 or D4+? and function below.

File: engines/director/lingo/lingo-bytecode.cpp (hunk 4 above)
Comment (moralrecordings) [reply engines/director/lingo/lingo-bytecode.cpp]: c_unk/c_unk1/c_unk2 will print out warning traces when the missing instructions are executed, is it worthwhile adding more for the loader?

File: engines/director/lingo/lingo-bytecode.cpp (hunk 5 above)
Comment (moralrecordings) [reply engines/director/lingo/lingo-bytecode.cpp]: Yep, sometimes e.g. two bytes are expected to be interpreted a single uint16.

File: engines/director/lingo/lingo-codegen.cpp (hunk 10 above)
Comment (moralrecordings) [reply engines/director/lingo/lingo-codegen.cpp]: Originally codeConst() would push a c_constpush instruction and then an encoded integer to the current script. As far as I could tell there was no shortcut for pushing only the encoded integer to the script, something which needs to be done quite a lot, so I adjusted codeConst to not add the push instruction, similar to how the other codeX shortcuts work. I re-added the push instruction to the few places codeConst was originally called. I've still yet to go and replace all of the longhand extra WRITE_UINT32 calls in lingo-gr.y with codeInt()

File: engines/director/lingo/lingo-code.cpp (hunk 8 above)
Comment (moralrecordings) [reply engines/director/lingo/lingo-code.cpp]: The naming was slightly confusing; Lingo scripts have a constants table and there are opcodes which push objects from that table using indexes, as distinct from pushing an integer/float encoded in the program listing.

File: engines/director/lingo/lingo-gr.y (hunk 11 above)
Comment (moralrecordings) [reply engines/director/lingo/lingo-gr.y]: See codeConst() answer

File: engines/director/score.cpp (hunk 12 above)
Comment (moralrecordings) [reply engines/director/score.cpp]: Not sure! I haven't tried taking apart a D5 game yet

File: engines/director/lingo/lingo-code.cpp (hunk 9 above)
Comment (moralrecordings) [reply engines/director/lingo/lingo-code.cpp]: Fixed

File: engines/director/lingo/lingo-bytecode.cpp (hunk 2 above)
Comment (moralrecordings) [reply engines/director/lingo/lingo-bytecode.cpp]: Fixed

File: engines/director/lingo/lingo-bytecode.cpp (hunk 1 above)
Comment (moralrecordings) [reply engines/director/lingo/lingo-bytecode.cpp]: Fixed

File: engines/director/lingo/lingo-bytecode.cpp (hunk 3 above)
Comment (moralrecordings) [reply engines/director/lingo/lingo-bytecode.cpp]: Fixed

File: engines/director/lingo/lingo-code.cpp (hunk 7 above)
Comment (moralrecordings) [reply engines/director/lingo/lingo-code.cpp]: I think that was a paste mistake, removed

File: engines/director/score.cpp (hunk 12 above)
Comment (moralrecordings) [reply engines/director/score.cpp]: Nice, D5 games share most of the data structures from D4, will open it up.

File: engines/director/score.cpp (hunk 13 above)
Comment (moralrecordings) [reply engines/director/score.cpp]: Fixed

File: engines/director/lingo/lingo-bytecode.cpp (hunk 6 above)
Comment (moralrecordings) [reply engines/director/lingo/lingo-bytecode.cpp]: Done

File: engines/director/lingo/lingo-bytecode.cpp (hunk 4 above)
Comment (moralrecordings) [reply engines/director/lingo/lingo-bytecode.cpp]: Added

File: engines/director/lingo/lingo-bytecode.cpp (hunk 6 above)
Comment (sev-) [reply engines/director/lingo/lingo-bytecode.cpp]: I think there is a misunderstanding. What I meant here is to do the following:
```unk1 = stream.readUint16();
unk2 = stream.readUint16();
...
debug(8, "funcitionentry: nameIndex=%d, unk1=%d, length=%d ...", nameIndex, unk1, length, unk2, unk3...);

**Comment by sev-** [comment]
Looks great, merging. You will be able to add that debug statement later, or I'll do it.
